src/shared/openssl-util.c

```diff
@@ -1298,6 +1298,105 @@ int pkey_generate_volume_keys(
+
+        _cleanup_(OSSL_STORE_closep) OSSL_STORE_CTX *store = OSSL_STORE_open(
+                        private_key_uri,
+                        /* ui_method= */ NULL,
+                        /* ui_data= */ NULL,
+                        /* post_process= */ NULL,
+                        /* post_process_data= */ NULL);
+        if (!store)
+                return log_openssl_errors("Failed to open OpenSSL store via '%s'", private_key_uri);
+
+        _cleanup_(OSSL_STORE_INFO_freep) OSSL_STORE_INFO *info = OSSL_STORE_load(store);
+        if (!info)
+                return log_openssl_errors("Failed to load OpenSSL store via '%s'", private_key_uri);
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *private_key = OSSL_STORE_INFO_get1_PKEY(info);
+        if (!private_key)
+                return log_openssl_errors("Failed to load private key via '%s'", private_key_uri);
+
+        EVP_PKEY_free(*ret_private_key);
+        *ret_private_key = TAKE_PTR(private_key);
+
+        return 0;
+#else
+        return -EOPNOTSUPP;
+#endif
+}
+
+DISABLE_WARNING_DEPRECATED_DECLARATIONS
+DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(ENGINE*, ENGINE_free, NULL);
+REENABLE_WARNING
+
+static int load_key_from_engine(
+                const char *engine,
+                const char *private_key_uri,
+                EVP_PKEY **ret_private_key) {
+
+        assert(engine);
+        assert(private_key_uri);
+        assert(ret_private_key);
+
+        DISABLE_WARNING_DEPRECATED_DECLARATIONS
+        _cleanup_(ENGINE_freep) ENGINE *e = ENGINE_by_id(engine);
+        if (!e)
+                return log_openssl_errors("Failed to load signing engine '%s'", engine);
+
+        if (ENGINE_init(e) == 0)
+                return log_openssl_errors("Failed to initialize signing engine '%s'", engine);
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *private_key = ENGINE_load_private_key(
+                        e,
+                        private_key_uri,
+                        /* ui_method= */ NULL,
+                        /* callback_data= */ NULL);
+        if (!private_key)
+                return log_openssl_errors("Failed to load private key from '%s'", private_key_uri);
+        REENABLE_WARNING
+
+        EVP_PKEY_free(*ret_private_key);
+        *ret_private_key = TAKE_PTR(private_key);
+
+        return 0;
+}
+
+int openssl_load_key_from_provider(const char *private_key_uri, EVP_PKEY **ret_private_key) {
+        _cleanup_free_ char *provider = NULL;
+        char *colon;
+        int r;
+
+        assert(private_key_uri);
+
+        colon = strchr(private_key_uri, ':');
+        if (!colon)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid URI '%s'", private_key_uri);
+
+        provider = strndup(private_key_uri, colon - private_key_uri);
+        if (!provider)
+                return log_oom_debug();
+
+        r = load_key_from_provider(provider, private_key_uri, ret_private_key);
+        if (r < 0)
```

Comment:
Unconditionally falling back seems a bit icky, but I don't really see a better solution unless we can figure out in which cases it makes sense to fall back and in which cases it doesn't.

Maybe add an environment variable to force either usage of providers or engines and if not set, we do the current logic of trying provider first and engine as a fallback?

Reply:
I can add an env var - when I have no provider, with my current version of openssl it doesn't fail to open the provider even if there isn't any, it only fails later when trying to load the key, so yeah it needs to be a generic fallback for it to work reliably.

Reply:
We should debug log here that we're falling back to engines